File: src/zig/gfx/include/brush.zig
const trans = @import("trans.zig");

const Transform = trans.Transform;

pub const Brush = extern struct {
    id: u64,
};

pub const Condition = extern struct {
    id: u64,
};

// Create brushes
pub extern fn brush(name: []const u8) *Brush;

// Apply a material when a condition is met
pub extern fn when(brush: *Brush, condition: *Condition, material: u16) *Brush;

// Combine brushes (later brushes override earlier ones)
pub extern fn layer(brushes: []const Brush) *Brush;

// Create conditions
pub extern fn depth(min: f32, max: f32) *Condition;
pub extern fn height(min: f32, max: f32) *Condition;
pub extern fn slope(min: f32, max: f32) *Condition;
pub extern fn noise(seed: u64, threshold: f32, scale: f32) *Condition;
pub extern fn curvature(min: f32, max: f32) *Condition;
pub extern fn distance(point_x: f32, point_y: f32, point_z: f32, min: f32, max: f32) *Condition;

// Logical operators for conditions
pub extern fn logical_and(a: *Condition, b: *Condition) *Condition;
pub extern fn logical_or(a: *Condition, b: *Condition) *Condition;
pub extern fn logical_not(condition: *Condition) *Condition;

// Special brush operations
pub extern fn scatter(base_brush: *Brush, feature_brush: *Brush, density: f32, seed: u64) *Brush;

// Apply brushes to an SDF to generate material data
pub extern fn paint(sdf: *anyopaque, size_x: u32, size_y: u32, size_z: u32, brush: *Brush) ?*anyopaque;

File: src/gfx/src/macos/Brush.swift
import Foundation

@frozen public struct Brush {
  let id: UInt64
  public init() {
    self.id = 0
  }
}

@frozen public struct Condition {
  let id: UInt64
  public init() {
    self.id = 0
  }
}

// MARK: - C-Callable Functions
@_cdecl("brush")
public func brush(name: UnsafeRawPointer?) -> UnsafeMutableRawPointer? {
  print("brush (Swift stub)")
  return nil
}

@_cdecl("when")
public func when(
  brush: UnsafeMutableRawPointer?, condition: UnsafeMutableRawPointer?, material: UInt16
) -> UnsafeMutableRawPointer? {
  print("when (Swift stub)")
  return nil
}

@_cdecl("layer")
public func layer(brushes: UnsafeMutableRawPointer?, count: Int32) -> UnsafeMutableRawPointer? {
  print("layer (Swift stub)")
  return nil
}

@_cdecl("depth")
public func depth(min: Float, max: Float) -> UnsafeMutableRawPointer? {
  print("depth (Swift stub)")
  return nil
}

@_cdecl("height")
public func height(min: Float, max: Float) -> UnsafeMutableRawPointer? {
  print("height (Swift stub)")
  return nil
}

@_cdecl("slope")
public func slope(min: Float, max: Float) -> UnsafeMutableRawPointer? {
  print("slope (Swift stub)")
  return nil
}

@_cdecl("noise")
public func noise(seed: UInt64, threshold: Float, scale: Float) -> UnsafeMutableRawPointer? {
  print("noise (Swift stub)")
  return nil
}

@_cdecl("curvature")
public func curvature(min: Float, max: Float) -> UnsafeMutableRawPointer? {
  print("curvature (Swift stub)")
  return nil
}

@_cdecl("distance")
public func distance(point_x: Float, point_y: Float, point_z: Float, min: Float, max: Float)
  -> UnsafeMutableRawPointer?
{
  print("distance (Swift stub)")
  return nil
}

@_cdecl("logical_and")
public func logical_and(a: UnsafeMutableRawPointer?, b: UnsafeMutableRawPointer?)
  -> UnsafeMutableRawPointer?
{
  print("logical_and (Swift stub)")
  return nil
}

@_cdecl("logical_or")
public func logical_or(a: UnsafeMutableRawPointer?, b: UnsafeMutableRawPointer?)
  -> UnsafeMutableRawPointer?
{
  print("logical_or (Swift stub)")
  return nil
}

@_cdecl("logical_not")
public func logical_not(condition: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
  print("logical_not (Swift stub)")
  return nil
}

@_cdecl("scatter")
public func scatter(
  base_brush: UnsafeMutableRawPointer?, feature_brush: UnsafeMutableRawPointer?, density: Float,
  seed: UInt64
) -> UnsafeMutableRawPointer? {
  print("scatter (Swift stub)")
  return nil
}

@_cdecl("paint")
public func paint(
  sdf: UnsafeMutableRawPointer?, size_x: UInt32, size_y: UInt32, size_z: UInt32,
  brush: UnsafeMutableRawPointer?
) -> UnsafeMutableRawPointer? {
  print("paint (Swift stub)")
  return nil
}
